keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -0,0 +1,256 @@
+  - [Pruning for CustomResources KEP](https://github.com/kubernetes/enhancements/pull/709)
+---
+
+# Title
+
+Graduate CustomResourceDefinitions to GA
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Defaulting and pruning for custom resources is implemented](#defaulting-and-pruning-for-custom-resources-is-implemented)
+  * [CRD v1 schemas are restricted to a subset of the OpenAPI specification](#crd-v1-schemas-are-restricted-to-a-subset-of-the-openapi-specification)
+  * [Generator exists for CRD Validation Schema v3 (Kubebuilder)](#generator-exists-for-crd-validation-schema-v3-kubebuilder)
+  * [CustomResourceWebhookConversion API is GA ready](#customresourcewebhookconversion-api-is-ga-ready)
+  * [CustomResourceSubresources API is GA ready](#customresourcesubresources-api-is-ga-ready)
+  * [v1 API](#v1-api)
+* [Test Plan](#test-plan)
+  * [Integration tests for GA](#integration-tests-for-ga)
+  * [e2e tests for GA](#e2e-tests-for-ga)
+  * [Conformance plan for GA](#conformance-plan-for-ga)
+  * [Scale Targets for GA](#scale-targets-for-ga)
+* [Graduation Criteria](#graduation-criteria)
+* [Post-GA tasks](#post-ga-tasks)
+  * [Arbitrary subresources as Post-GA task](#arbitrary-subresources-as-post-ga-task)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+CustomResourceDefinitions (CRDs) are the way to extend the Kubernetes API to
+include custom resource types that behave like the native resource types. CRDs
+have been in Beta since Kubernetes 1.7. This document outlines the required
+steps to graduate CRDs to general availability (GA).
+
+## Motivation
+
+CRDs are in widespread use as a Kubernetes extensibility mechanism and have been
+available in beta for the last 5 Kubernetes releases. Feature requests and bug
+reports suggest CRDs are nearing GA quality, and this KEP aims to itemize the
+remaining improvements to move this toward GA.
+
+### Goals
+
+Establish consensus for the remaining essential improvements needed to move CRDs to GA.
+
+Guiding principles:
+* if a missing feature forces or encourages users to implement non-Kubernetes-like APIs, and therefore damages the ecosystem long term, it belongs on this list.
+* If a feature cannot be added as an after-thought of a GA API, it or some preparation belongs on this list.
+
+
+### Non-Goals
+
+Full feature parity with native kubernetes resource types is not a GA graduation goal. See above guiding principles and below Post-GA tasks.
+
+## Proposal
+
+Objectives to graduate CRDs to GA, each of which is described in more detail further below:
+
+* Defaulting and pruning for custom resources is implemented
+* CRD v1 schemas are restricted to a subset of the OpenAPI specification (and there is a v1beta1 compatibility plan)
+* Generator exists for CRD Validation Schema v3 (Kubebuilder)
```

Comment:
more precisely it is crd-gen in sigs.k8s.io/controller-tools, used by kubebuilder.

Reply:
nvmd. Details are below.